src/expr/src/scalar/mod.rs

```diff
@@ -766,6 +766,103 @@ impl MirScalarExpr {
+                            let top_level_list_create = &mut exprs[0];
+                            let mut list_creates = vec![top_level_list_create];
+                            let mut i = 0;
+                            let mut all_indexes_removed = true;
+                            while i < index_exprs.len() {
+                                if !list_creates.iter().all(|lc| is_list_create_call(lc)) {
+                                    all_indexes_removed = false;
+                                    break;
+                                }
+                                if index_exprs[i].is_literal_ok() {
+                                    // We can remove this index.
+                                    let removed_index = index_exprs[i].clone();
+                                    index_exprs.remove(i);
+                                    if let Datum::Int64(sql_index_i64) =
+                                        removed_index.as_literal().unwrap().unwrap()
+                                    {
+                                        let index_i64 = sql_index_i64 - 1;
+                                        // substitute each list_create in list_creates by its
+                                        // `index`th argument (or null)
+                                        for list_create in &mut list_creates {
+                                            if let MirScalarExpr::CallVariadic {
+                                                func: VariadicFunc::ListCreate { .. },
+                                                exprs: list_create_args,
+                                            } = list_create
+                                            {
+                                                // ListIndex gives null on an out-of-bounds index
+                                                if index_i64 >= 0
+                                                    && index_i64
+                                                        < list_create_args.len().try_into().unwrap()
+                                                {
+                                                    let index: usize =
+                                                        index_i64.try_into().unwrap();
+                                                    **list_create =
+                                                        list_create_args.swap_remove(index);
+                                                } else {
+                                                    let typ = list_create_type(list_create);
+                                                    **list_create =
+                                                        MirScalarExpr::literal_null(typ);
+                                                }
+                                            } else {
+                                                // func cannot be anything else than a ListCreate
+                                                unreachable!()
```

Comment:
How about making these early panics, to keep the indentation level in check?

Also, you could consider making this whole optimization a separate function. I think the code size warrants this.

Reply:
Both done. When moving the code to a separate function, I had to get `exprs` by a match again at the beginning of the function. At first, I tried passing both `e` and `exprs` as an argument, but then the compiler was complaining that these are two mutable borrows. If I get `exprs` at the beginning of the function, then I think it works because the compiler can see that `e` is used only at the end after all the uses of `exprs`.

Reply:
If you'd wanted to get rid of that extra unwrap what I think would work is if you returned the optimized list_index expression, instead of modifying it in-place. E.g.:

```diff
diff a/src/expr/src/scalar/mod.rs b/src/expr/src/scalar/mod.rs
@@ -770,7 +770,8 @@ impl MirScalarExpr {
                         {
                             // We are looking for list_index(list_create, literal), and eliminate
                             // both the list_index and the list_create. E.g.: LIST[f1,f2][2] --> f2
-                            list_create_list_index_literal(e);
+                            let exprs = mem::take(exprs);
+                            *e = list_create_list_index_literal(exprs);
                         }
                     }
                     MirScalarExpr::If { cond, then, els } => {
@@ -896,7 +897,7 @@ impl MirScalarExpr {
             },
         );

-        fn list_create_list_index_literal(e: &mut MirScalarExpr) {
+        fn list_create_list_index_literal(mut exprs: Vec<MirScalarExpr>) -> MirScalarExpr {
             // We handle multi-dimensional lists as well, e.g.
             // LIST[[[f1, f2], [f3, f4]], [[f5, f6], [f7, f8]]] [2][n][2] --> LIST[f6, f8] [n]
             // We iterate over the index_exprs and remove literals, but keep non-literals.
@@ -906,10 +907,6 @@ impl MirScalarExpr {
             // When we remove a literal, we need to partial-evaluate all ListCreates
             // that are at the current level (except those that disappeared due to
             // literals at earlier levels).
-            let exprs = match e {
-                MirScalarExpr::CallVariadic { func: _, exprs } => exprs,
-                _ => unreachable!(), // we already matched it before calling the function
-            };
             let index_exprs = &mut exprs.split_off(1);
             let top_level_list_create = &mut exprs[0];
             let mut list_creates = vec![top_level_list_create];
@@ -970,12 +967,16 @@ impl MirScalarExpr {
                 assert_eq!(list_creates.len(), 1);
                 // Remove the ListIndex, because it would have 0 arguments.
                 // (list_creates[0] is not necessarily a ListCreate at this point.)
-                *e = list_creates[0].take();
+                list_creates[0].take()
             } else {
                 // Either some of the indexes were not literals, or we broke out from the loop.
                 // Put back the remaining index expressions. (split_off removed them
                 // at the beginning.)
                 exprs.append(index_exprs);
+                MirScalarExpr::CallVariadic {
+                    func: VariadicFunc::ListIndex,
+                    exprs,
+                }
             }
         }
```

Reply:
Done, thank you!